### ISSUE TITLE
bug during reset value of a string array

### ISSUE THREAD

**Issue by jiayi797** [numpy/numpy#13418]
# Code

```python
import numpy as np

a = np.array(['11','22']) 
print(a) # output ['11' '22']
a[0] = '121'
print(a) # output ['12' '22'] !!!!
```

when I want to reset a[0] as a longer string, it will cut off my new value. it cause a bug in my system and it takes my whole week. >_<

I rewrite it to :
```python
a = np.array(['11','22'],dtype = 'U10') 
print(a) # output ['11' '22']
a[0] = '121'
print(a) # output ['121' '22'] 
```

it seems work well. While I think it might cause confused again in future.

**Comment by seberg** [comment]
I tend to agree, similar to what Complex does. Although, we do currently use unsafe casting for assignments (i.e. also floats assigned to integer arrays get truncated). That is a bit less nasty though, so may be worth to specialize this a  bit.

**Comment by jiayi797** [comment]
> I tend to agree, similar to what Complex does. Although, we do currently use unsafe casting for assignments (i.e. also floats assigned to integer arrays get truncated). That is a bit less nasty though, so may be worth to specialize this a bit.

unsafe casting from float to integer is common, such as java/c/c++/c#.
while I think string unsafe casting is a dangerous operator. Support that sting is some key identification, and most developer don't think string have unsafe casting. I think it will cause much confused .
